test/results_test.dart

```diff
@@ -741,4 +741,42 @@ Future<void> main([List<String>? args]) async {
     final items = realm.query<Product>(ids);
     expect(items.length, 0);
   });
+
+  test('RealmResults.first throws on empty', () async {
+    final config = Configuration.local([Dog.schema, Person.schema]);
+    final realm = getRealm(config);
+
+    final empty = Iterable<int>.empty();
+    expect(() => empty.first, throwsA(isA<StateError>()));
+    expect(() => realm.all<Dog>().first, throwsA(isA<RealmException>()));
+  });
+
+  test('RealmResults.last throws on empty', () async {
+    final config = Configuration.local([Dog.schema, Person.schema]);
+    final realm = getRealm(config);
+
+    final empty = Iterable<int>.empty();
+    expect(() => empty.last, throwsA(isA<StateError>()));
+    expect(() => realm.all<Dog>().last, throwsA(isA<RealmException>()));
```

Comment:
These two lines show the discrepancy. If we try to mimic the state Iterable then `last` should throw 
`Bad state: No element` for Realm as well. At least it should be `RealmStateError` when empty

Reply:
I believe `first` as well

Reply:
A reason not to do that is that it will require an extra call to native to check the `length`, and another error is already raised if not. But if we deem the extra call acceptable, then I can easily add it like in `single` (where it is needed to raise some error at all)

Reply:
Latest commit test raises `RealmStateError` on `first` and `last` as well, if `RealmResults` is empty